File upload with metadata

jmozah opened the pull request:
This PR should do the following
- Create  POST http://host:port/bzz-file api to upload a file
- Create GET http://host:port/bzz-file/{address} to download a file with the filename uploaded

See the user story for more details https://app.zenhub.com/workspaces/swarm-10-5e7c6d9966b97049231c3c0b/issues/ethersphere/bee/274

jmozah commented:
@janos  @nolash  I have done most of the part except testing upload with multiple files. If you insist i will do that, but since that was not a requirement i am submitting for a review. Also the Coverage report failing looks like a false alarm.

nolash commented on pkg/api/bzzfile.go:
```diff
@@ -0,0 +1,214 @@
+
+func (s *server) bzzFileUploadHandler(w http.ResponseWriter, r *http.Request) {
+	contentType, params, err := mime.ParseMediaType(r.Header.Get("Content-Type"))
+	if contentType != MultiPartFormData {
```
Support for application/octet-stream upload will be added in later PR?

jmozah replied:
I will add this in a new PR @nolash

nolash commented on pkg/api/bzzfile.go:
```diff
@@ -0,0 +1,214 @@
+			jsonhttp.BadRequest(w, "error parsing parse mediatype")
+			return
+		}
+		fileName := params["name"]
```
I see there is a method called `FileName` in `mime/multipart`. Would that be useful here?

nolash commented on pkg/api/bzzfile.go:
```diff
@@ -0,0 +1,214 @@
+		}
+		fileName := params["name"]
+		if fileName == "" {
+			fileName = part.FormName()
```
```
func (p *Part) FormName() string
    FormName returns the name parameter if p has a Content-Disposition of type
    "form-data". Otherwise it returns the empty string.
```

It seems the filename can still be an empty string after this. The `bee-file` binary sets the filename to the hex string of the swarm address when no other filename has been set. Should we aim for the same behavior here?

nolash commented on pkg/api/bzzfile.go:
```diff
@@ -0,0 +1,214 @@
+		ctx := context.Background()
+
+		// first store the file and get the reference
+		buf, err := ioutil.ReadAll(part)
```
Isn't it better to pass this directly to the splitter instead of this extra buffering?

nolash commented on pkg/api/bzzfile.go:
```diff
@@ -0,0 +1,214 @@
+			jsonhttp.BadRequest(w, "read a part data")
+			return
+		}
+		fr, err := s.storeData(ctx, splitr, buf)
```
I prepared the `pkg/api/raw.go:splitUpload` so it would be re-usable. Was it not good enough since you decided to write new code for this?

nolash commented on pkg/api/bzzfile.go:
```diff
@@ -0,0 +1,214 @@
+		m := entry.NewMetadata(fileName)
+		contentType := part.Header.Get("Content-Type")
+		if contentType == "" {
+			contentType = http.DetectContentType(buf)
```
This magic is the reason why you do the extra buffer? I am (still) of the opinion that this kind of magic is unnecessary for a lean bee node.

nolash commented on pkg/api/bzzfile.go:
```diff
@@ -0,0 +1,214 @@
+		mr, err := s.storeData(ctx, splitr, metadataBytes)
+		if err != nil {
+			s.Logger.Debugf("bzz-file: metadata store: %v, file name %s", err, fileName)
+			s.Logger.Error("bzz-file: metadata store ")
```
space

nolash replied:
still space here

nolash commented on pkg/api/bzzfile.go:
```diff
@@ -0,0 +1,214 @@
+	Address swarm.Address `json:"address"`
+}
+
+func (s *server) bzzFileUploadHandler(w http.ResponseWriter, r *http.Request) {
```
please describe the methods

nolash commented on pkg/api/bzzfile.go:
```diff
@@ -0,0 +1,214 @@
+		return
+	}
+	w.Header().Set("ETag", fmt.Sprintf("%q", e.Reference()))
+	w.Header().Set("Content-Disposition", fmt.Sprintf("inline; filename=\"%s\"", metaData.Filename))
```
Would it be an idea to control through request query whether disposition should be inline or attachment?

jmozah replied:
Good thought. For now i am hard coding it so that the browser can show it. later we can make it based on the query.

nolash commented on pkg/api/bzzfile.go:
```diff
@@ -0,0 +1,214 @@
+	dataReadCloser := ioutil.NopCloser(dataReader)
+	dataAddr, err := splitr.Split(ctx, dataReadCloser, int64(len(dataBytes)))
+	if err != nil {
+		return swarm.NewAddress([]byte{0}), err
```
Please use `swarm.ZeroAddress`

nolash commented on pkg/api/bzzfile.go:
```diff
@@ -0,0 +1,214 @@
+)
+
+type FileUploadResponse struct {
+	Address swarm.Address `json:"address"`
```
I think it should be named `Reference` rather than `Address`

nolash commented on pkg/api/bzzfile_test.go:
```diff
@@ -0,0 +1,107 @@
+	})
+
+	t.Run("upload-then-download-and-check-data", func(t *testing.T) {
+		fileName := "vacation.mp4"
```
mp4 but html content? I don't understand..

jmozah replied:
well.. the detection is based on content.. filename makes no difference.. for easy reading i am making it sample.html :-)

nolash commented on pkg/api/raw.go:
```diff
@@ -66,9 +66,9 @@ func (s *server) splitUpload(ctx context.Context, r io.ReadCloser, l int64) (int
 		response.Code = http.StatusInternalServerError
 		err = fmt.Errorf("%s: %v", response.Message, err)
-		return response, err
+		return swarm.NewAddress([]byte{0}), err
```
Please use `swarm.ZeroAddress`

nolash commented on pkg/collection/entry/metadata.go:
```diff
@@ -21,6 +21,10 @@ func NewMetadata(fileName string) *Metadata {
 	}
 }
 
+func (m *Metadata) SetMimeType(mineType string) {
```
IIRC I was already reprimanded for adding this setter to a simple exported type.

nolash commented on pkg/jsonhttp/jsonhttptest/jsonhttptest.go:
```diff
@@ -35,6 +39,80 @@ func ResponseDirect(t *testing.T, client *http.Client, method, url string, body
 
+func ResponseDirectWithMultiPart(t *testing.T, client *http.Client, method, url, fileName string, data []byte,
+	responseCode int, response interface{}) http.Header {
+	body := &bytes.Buffer{}
```
`NewBuffer`?

nolash commented on pkg/jsonhttp/jsonhttptest/jsonhttptest.go:
```diff
@@ -35,6 +39,80 @@ func ResponseDirect(t *testing.T, client *http.Client, method, url string, body
+	if err != nil {
+		t.Error(err)
+	}
+	_, err = io.Copy(part, ioutil.NopCloser(bytes.NewReader(data)))
```
Why do you need the `NopCloser` enhancement here?

nolash commented on pkg/jsonhttp/jsonhttptest/jsonhttptest.go:
```diff
@@ -35,6 +39,80 @@ func ResponseDirect(t *testing.T, client *http.Client, method, url string, body
+	if err != nil {
+		t.Error(err)
+	}
+	//req.Header.Set("Expect", "100-continue")
```
Please remove commented code

janos commented:
@jmozah could you update this branch with the recent changes from the master? We have update the API router.

janos commented on pkg/api/router.go:
```diff
@@ -28,6 +28,14 @@ func (s *server) setupRouting() {
 		fmt.Fprintln(w, "User-agent: *\nDisallow: /")
 	})
 
+	router.Handle("/bzz-file", jsonhttp.MethodHandler{
+		"POST": http.HandlerFunc(s.bzzFileUploadHandler),
+	})
+
+	router.Handle("/bzz-file/{addr}", jsonhttp.MethodHandler{
+		"GET": http.HandlerFunc(s.bzzFileDownloadHandler),
+	})
+
```
Could you update this branch to the latest master and use the handle function from there as:

```
	handle(router, "/files", jsonhttp.MethodHandler{
		"POST": http.HandlerFunc(s.fileUploadHandler),
	})
	handle(router, "/files/{address}", jsonhttp.MethodHandler{
		"GET": http.HandlerFunc(s.fileGetHandler),
	})
```

jmozah commented:
> @jmozah could you update this branch with the recent changes from the master? We have updated the API router.

@janos Updated and fixed the review comments

janos commented on pkg/api/bytes.go:
```diff
@@ -66,7 +66,7 @@ func (s *server) splitUpload(ctx context.Context, r io.ReadCloser, l int64) (int
 		response.Code = http.StatusInternalServerError
 		err = fmt.Errorf("%s: %v", response.Message, err)
-		return response, err
+		return swarm.ZeroAddress, err
```
The `response` variable is not used, it can be removed, also the fmt.Errorf.

jmozah replied:
moved this up to the `bytesUploadHandler` for now

janos commented on pkg/api/bzzfile.go:
```diff
@@ -0,0 +1,245 @@
+			return
+		}
+
+		ctx := context.Background()
```
Use r.Context().

janos commented on pkg/api/bzzfile.go:
```diff
@@ -0,0 +1,245 @@
+		// then find out content type
+		contentType := part.Header.Get("Content-Type")
+		if contentType == "" {
+			reader := bufio.NewReader(part)
```
`part` is already read and here cannot be read any further as it reached EOF. The whole content type detection has to be done before any Read call on `part`. Basically, before storePartData call, the whole code block from line 85 to line 111 has to be moved, with one important difference. If the bufio.Reader is constructed it has to be passed to storePartData, not directly the `part` variable as `part` reader already moved its cursor, while the bufio reader did not, it just peeked, so it can read from the start, as `part` reader has advanced 512 bytes. The best way to do that is to create `var reader io.Reader = part` before `if contentType == "" {` line and here to do `reader = bufio.NewReader(part)` (assign to the variable, not create a new one), and use that `reader` variable in storePartData `fr, err := s.storePartData(ctx, reader, fileSize)`.

It looks to me that `contentType == ""` block is not tested and it would be good to have that test.

janos commented on pkg/api/bzzfile.go:
```diff
@@ -0,0 +1,245 @@
+	}
+
+	// send the file data back in the response
+	outBuffer := bytes.NewBuffer(nil)
+	_, err = file.JoinReadAll(j, e.Reference(), outBuffer)
+	if err != nil {
+		s.Logger.Debugf("bzz-file: data read %s: %v", addr, err)
+		s.Logger.Error("bzz-file: data read")
+		jsonhttp.InternalServerError(w, "error reading data")
+		return
+	}
```
This whole block has the same problem as JoinReadAll described in https://github.com/ethersphere/bee/pull/266#discussion_r437311701. At least the jsonhtt call should be removed and we can handle the err != nil before any write to outBuffer in another PR. BYW, I think that outBuffer does not have any purpose in the current code.

janos commented on pkg/api/bzzfile.go:
```diff
@@ -0,0 +1,245 @@
+// storeMeta is used to store metadata information as a whole.
+func (s *server) storeMeta(ctx context.Context, dataBytes []byte) (swarm.Address, error) {
+	dataBuf := bytes.NewBuffer(dataBytes)
+	dataReader := io.LimitReader(dataBuf, int64(len(dataBytes)))
```
I do not think that LimitReader is needed. It limits the rad to exact how the data is long.

jmozah replied:
I use it only in meta writing.. where we exactly know how much bytes to write.. is it not safe to have it?

janos commented on pkg/api/bzzfile.go:
```diff
@@ -0,0 +1,245 @@
+	if err != nil {
+		return swarm.ZeroAddress, err
+	}
+	bytesResp := o.(bytesPostResponse)
```
This is very unfortunate. But this is a consequence of a badly written splitUpload. We can address this later in the PR when we address technical dept created in the pr that added splitUpload.

janos commented on pkg/api/router.go:
```diff
@@ -38,6 +38,15 @@ func (s *server) setupRouting() {
 		"POST": http.HandlerFunc(s.bytesUploadHandler),
 	})
+
+	handle(router, "/bzz-file", jsonhttp.MethodHandler{
+		"POST": http.HandlerFunc(s.bzzFileUploadHandler),
+	})
+
+	handle(router, "/bzz-file/{addr}", jsonhttp.MethodHandler{
```
We can call this endpoint `/files` right away as per https://hackmd.io/fGoeV1n2QficZwuHnYmb-g#File.

jmozah commented:
> part is already read and here cannot be read any further as it reached EOF. The whole content type detection has to be done before any Read call on part. Basically, before storePartData call, the whole code block from line 85 to line 111 has to be moved, with one important difference. If the bufio.Reader is constructed it has to be passed to storePartData, not directly the part variable as part reader already moved its cursor, while the bufio reader did not, it just peeked, so it can read from the start, as part reader has advanced 512 bytes. The best way to do that is to create var reader io.Reader = part before if contentType == "" { line and here to do reader = bufio.NewReader(part) (assign to the variable, not create a new one), and use that reader variable in storePartData fr, err := s.storePartData(ctx, reader, fileSize).

Strange, i dont know how the unit testcase for this worked which also test and check the downloaded data with uploaded data. but i changed it as you said.

> It looks to me that contentType == "" block is not tested and it would be good to have that test.

Added arguments in test to cover the scenario where we dont send contentType, and check it on receiving.

janos commented on pkg/api/router.go:
```diff
@@ -38,6 +38,15 @@ func (s *server) setupRouting() {
 		"POST": http.HandlerFunc(s.bytesUploadHandler),
 	})
+
+	handle(router, "/file", jsonhttp.MethodHandler{
+		"POST": http.HandlerFunc(s.bzzFileUploadHandler),
+	})
+
+	handle(router, "/file/{addr}", jsonhttp.MethodHandler{
+		"GET": http.HandlerFunc(s.bzzFileDownloadHandler),
+	})
```
The endpoint should be in plural `/files`.

janos commented on pkg/api/file.go:
```diff
@@ -0,0 +1,265 @@
+// storeMeta is used to store metadata information as a whole.
+func (s *server) storeMeta(ctx context.Context, dataBytes []byte) (swarm.Address, error) {
+	dataBuf := bytes.NewBuffer(dataBytes)
+	dataReader := io.LimitReader(dataBuf, int64(len(dataBytes)))
```
LimitReader is not needed.

nolash replied:
Just make sure that splitter writes aren't called with more than one chunk of data at a time...

janos commented on pkg/api/file.go:
```diff
@@ -0,0 +1,265 @@
+		// find file size
+		fileSizeString := part.Header.Get("Content-Length")
+		if fileSizeString == "" {
+			s.Logger.Debugf("file: content length: %v", err)
```
This error is nil here. The log message should be about missing content length.

Currently, I get no content length for a part when I submit with the form like this one:

```
<form action="http://localhost:8080/file" method="post" enctype="multipart/form-data">
 <div>
   <label for="file">Choose file to upload</label>
   <input type="file" id="file" name="file">
 </div>
 <div>
   <button>Submit</button>
 </div>
</form>
```

Do you maybe have an idea what is wrong? As I had this form working with the code from last week, which is not here anymore because of the force push.

janos replied:
I have created a pr that handles this like we do in swarm already https://github.com/ethersphere/bee/pull/296. With it, the upload is working nicely.

jmozah replied:
merged your PR
